verilog/analysis/flow_tree.h

```diff
@@ -0,0 +1,123 @@
+// Copyright 2017-2022 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_VERILOG_FLOW_TREE_H_
+#define VERIBLE_VERILOG_FLOW_TREE_H_
+
+#include <bitset>
+#include <map>
+#include <stack>
+#include <string>
+#include <vector>
+
+#include "absl/status/status.h"
+#include "common/lexer/token_stream_adapter.h"
+#include "verilog/parser/verilog_token_enum.h"
+
+namespace verilog {
+
+using BitSet = std::bitset<128>;
+using TokenSequenceConstIterator = verible::TokenSequence::const_iterator;
+
+struct ConditionalBlock {
+  TokenSequenceConstIterator ifdef_iterator;
+  TokenSequenceConstIterator ifndef_iterator;
+  std::vector<TokenSequenceConstIterator> elsif_iterators;
+  TokenSequenceConstIterator else_iterator;
+  TokenSequenceConstIterator endif_iterator;
+};
+
+struct Variant {
+  // Contains the token sequence of the variant.
+  verible::TokenSequence sequence;
+
+  // The i-th bit in "macros_mask" is 1 when the macro (with ID = i) is assumed
+  // to be defined, o.w. it is assumed to be undefined.
+  BitSet macros_mask;
+
+  // The i-th bit in "assumed" is 1 when the macro (with ID = i) is visited or
+  // assumed (either defined or not), o.w. it is not visited (its value doesn't
+  // affect this variant).
+  BitSet assumed;
```

Comment:
I don't really understand the description of these fields.
What is the difference of `macro_mask` and `assumed` ?

Reply:
`assumed` is used to identify if a macro's value doesn't matter a variant, e.g.

```verilog
`ifdef A
  `ifdef B
   ...
   `endif
`endif
```

In this case we have 2 macros, then each variant should be identified by a `bitset` (which is `macro_mask`) of length 2.
For example the `bitset macro_mask` of value `01`  shows the variant in which `A` is defined and `B` is not.
Similarly the `bitset macro_mask` of value `00` refers to the variant in which both `A` and `B` is undefined, but in fact when `A` is undefined, then `B` doesn't affect the variant and using `0` in `macro_mask` is not accurate enough because `10` will still refers to the same variant.

That's the use of `assumed` it is like another layer of information, if a bit in assumed is not set, then it means that this macro doesn't affect the variant in any way.

Reply:
It might be useful to add this description in the header file